Enable snapshot/restore functionalities

sboeuf opened the pull request:
This pull request finishes the series by enabling snapshot/restore functionality through the VM. It has limited scope for now as it does not support virtio and VFIO devices, but that's the first step in supporting snapshotting and restoring a VM.

sboeuf commented:
/!\ This PR depends on https://github.com/cloud-hypervisor/cloud-hypervisor/pull/995 /!\
Do NOT merge before https://github.com/cloud-hypervisor/cloud-hypervisor/pull/995 gets merged.

sboeuf commented:
One more thing, this PR needs an integration test that I will add tomorrow. But it is ready to be reviewed though.

sameo commented on vmm/src/cpu.rs:
```diff
@@ -823,6 +875,107 @@ impl CpuManager {
+                                error!("VCPU generated error: {:?}", e);
+                                break;
+                            }
+                            Ok(true) => {}
+                            Ok(false) => {
+                                reset_evt.write(1).unwrap();
+                                break;
+                            }
+                        }
+
+                        // We've been told to terminate
+                        if vcpu_kill_signalled.load(Ordering::SeqCst)
+                            || vcpu_kill.load(Ordering::SeqCst)
+                        {
+                            break;
+                        }
+
+                        // If we are being told to pause, we park the thread
+                        // until the pause boolean is toggled.
+                        // The resume operation is responsible for toggling
+                        // the boolean and unpark the thread.
+                        // We enter a loop because park() could spuriously
+                        // return. We will then park() again unless the
+                        // pause boolean has been toggled.
+                        while vcpu_pause_signalled.load(Ordering::SeqCst) {
+                            thread::park();
+                        }
+                    }
+                })
+                .map_err(Error::VcpuSpawn)?,
+        );
+
+        // On hot plug calls into this function entry_point is None. It is for
+        // those hotplug CPU additions that we need to set the inserting flag.
+        self.vcpu_states[usize::from(cpu_id)].handle = handle;
+        self.vcpu_states[usize::from(cpu_id)].inserting = false;
+
+        Ok(())
+    }
+
```
I have to blame myself for that one, and this was part of my TODO list, but there's a lot of common code between `restore_vcpu()` and the main `activate_vcpus()` loop. We have to factorize the vcpu thread creation code at least.

sboeuf replied:
Ok I'm gonna look at this!

sboeuf replied:
Done!

sboeuf commented:
>One more thing, this PR needs an integration test that I will add tomorrow. But it is ready to be reviewed though.

Ok so we decided we won't add an integration for now, as we want something simple relying on ssh. Once the snapshot/restore feature will be supported for virtio-net, we will be able to add this test.

sameo commented on vmm/src/cpu.rs:
```diff
@@ -545,6 +739,153 @@ impl CpuManager {
+                .configure(entry_point, &vm_memory, self.cpuid.clone())
+                .expect("Failed to configure vCPU");
+        }
+
```
Nice, thanks for that.

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -591,6 +591,7 @@ impl DeviceManager {
 
```
This commit message is too long:

`vmm: device_manger: Make serial and ioapic devices migratable`

sboeuf replied:
Ok!

rbradford commented on vmm/Cargo.toml:
```diff
@@ -19,7 +19,7 @@ anyhow = "1.0"
 arch = { path = "../arch" }
```
This commit message could do with a "why".

sboeuf replied:
I will update it!

rbradford commented on vmm/src/cpu.rs:
```diff
@@ -265,17 +265,17 @@ impl Vcpu {
         mmio_bus: Arc<devices::Bus>,
```
This commit message needs a "why"

sboeuf replied:
I will update it!

rbradford commented on vmm/src/cpu.rs:
```diff
@@ -384,6 +455,65 @@ impl Vcpu {
         );
     }
+
+    #[allow(unused)]
```
I don't see these being removed in a later commit.

sboeuf replied:
Yes it's removed here: https://github.com/cloud-hypervisor/cloud-hypervisor/pull/998/commits/a7c6560347f302fcab7826900c7e086f570f4b1f#diff-39805f31fede77dcaa77b9055d65e821L459

sboeuf commented:
@rbradford @sameo Ok I think I addressed your comments by expanding every commit message. Also, I implemented the copy of the memory region (filling the regions with the file content) instead of having the backing file.
As a follow up PR, I will look at supporting the VM template use case we've been talking about.

sameo commented on vmm/src/memory_manager.rs:
```diff
@@ -350,6 +353,66 @@ impl MemoryManager {
+                        return Err(Error::Restore(MigratableError::Restore(anyhow!(
+                            "Could not deserialize MemoryManager {}",
+                            error
+                        ))))
+                    }
+                };
+
+            memory_manager
+                .lock()
+                .unwrap()
+                .guest_memory
+                .memory()
+                .with_regions(|index, region| {
+                    let memory_region_path = PathBuf::from(format!(
+                        "{}/{}",
+                        vm_snapshot_path.display(),
+                        mem_snapshot.memory_regions[index].backing_file.display()
+                    ));
+
+                    // Create the snapshot file for the region
+                    let mut memory_region_file = OpenOptions::new()
+                        .read(true)
+                        .open(memory_region_path)
+                        .map_err(|e| Error::Restore(MigratableError::MigrateReceive(e.into())))?;
+
+                    region
+                        .read_from(
+                            MemoryRegionAddress(0),
+                            &mut memory_region_file,
+                            region.len().try_into().unwrap(),
+                        )
+                        .map_err(|e| Error::Restore(MigratableError::MigrateReceive(e.into())))?;
+
+                    Ok(())
+                })?;
+        }
+
+        Ok(memory_manager)
+    }
+
```
That looks simpler indeed, it's good.
Thanks for the refactoring, LGTM.

rbradford commented on vmm/src/memory_manager.rs:
```diff
@@ -342,6 +353,66 @@ impl MemoryManager {
+                .guest_memory
+                .memory()
+                .with_regions(|index, region| {
+                    let memory_region_path = PathBuf::from(format!(
```
Should really use the path functions for extending (does the right thing with extra/missing slashes). `PathBuf::push()` is what you want.

sboeuf replied:
Ok I'll update :)

sboeuf replied:
Done!
